Add common verification alignment

bradyrx opened the pull request:
# Description

This adds functionality for `alignment='same_verifs'`, following Boer et al. 2016. In short, this selects initializations at each lead that verify in the identical window over all leads. This is useful for bias correction in particular.

This will also take a big step toward refactoring and simplifying the way we handle hindcast computations.

Closes https://github.com/bradyrx/climpred/issues/318

## Type of change

Please delete options that are not relevant.

-   [X]  New feature (non-breaking change which adds functionality)
expected)
-   [X]  This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. This could point to a cell in the updated notebooks. Or a snippet of code with accompanying figures here.

## Checklist (while developing)

-   [x]  I have added docstrings to all new functions.
-   [x]  I have commented my code, particularly in hard-to-understand areas
-   [x]  Tests added for `pytest`, if necessary.
-   [x]  I have updated the sphinx documentation, if necessary.

## Pre-Merge Checklist (final steps)

-   [x]  I have rebased onto master or develop (wherever I am merging) and dealt with any conflicts.
-   [x]  I have squashed commits to a reasonable amount, and force-pushed the squashed commits.

## References

Please add any references to manuscripts, textbooks, etc.

bradyrx commented:
This sets us up for the refactoring framework I'm envisioning that can collapse all of the hindcast compute functions into one (including reference forecasts).

Typical `compute` workflow:

1. Run all checks at the header (e.g. `has_valid_lead_units`).
2. Retrieve metric/comparison.
3. Run alignment function. It now returns `inits` and `verif_dates`. This is all done prior to the looping through lead time.

`inits` is a dictionary with keys as the lead values and entries as a list of inits to select at each lead. E.g. `{1: [1990, 1991, 1992], 2: [1990, 1991, 1992], ...}` for `same_inits`.

`verif_dates` is a dictionary with keys as the lead values and entries as a list of dates to select at each lead for the verification product. E.g. `{1: [1991, 1992, 1993], 2: [1992, 1993, 1994]}`.

4. Loop through leads and select the inits from the forecast (`forecast.sel(lead=i, time=inits[i])`) and verification dates from the product (`verif.sel(time=verif_dates[i])`).
5. Run metric.
6. Post-processing on attributes, coordinates, etc.

bradyrx commented:
Some notes for a future PR...

How you'd use a reference in this framework. You'd have a keyword called `reference=...`. If `None` it just runs the initialized comparison. If `persistence`, both `forecast` and `verif` == `verif`. Then at step 4 above you still do `forecast.sel(time=inits[i])` and `verif.sel(time=verif_dates[i])`. (some light switches needed in there so not have `.sel(lead=i)`)

If `climatology`, you'd do `forecast = mean(verif)` over some defined time-period. Then maybe you broadcast it to look like the original forecast. Then you do `forecast.sel(time=inits[i])` (which is the same at all time, or maybe for monthly it has 12 values that repeat) and `verif.sel(time=verif_dates[i])`.

If `uninitialized` or `forced` or `control` it just takes in another time series or ensemble. Then you do `forecast.sel(verif_dates[i])` and `verif.sel(verif_dates[i])`.

Etc., etc. There'd just be function calls to retrieve these simple reference algorithms.

bradyrx commented:
@aaronspring, this is ready to go. The last one I'll do after this is to add the `maximize` keyword.

aaronspring commented on climpred/prediction.py:
```diff
@@ -276,12 +270,12 @@ def compute_hindcast(
-                f'inits={forecast["time"].min().values}'
-                f'-{forecast["time"].max().values} | '
+                f'inits={inits[i].min().values}'
+                f'-{inits[i].max().values} | '
```
didnt you want to add all times instead of min max only? or at least at the debug level?

bradyrx replied:
That will be done on the debug level in the next PR. I just want to get this one in so I can work on `maximize`. I'll put it on the list for maximize, though.

aaronspring commented on docs/source/images/alignment_plots/same_inits_alignment.ipynb:
```diff
@@ -0,0 +1,305 @@
+{
```
do you think you can put it all in one notebook?

bradyrx replied:
Yes, I'll merge these in the next PR.

aaronspring commented:
also this plotting requires proplot. is this necessary?

aaronspring commented:
I sent you a notebook in which I produced some breaking code. in the future we will need to adapt logging in bootstrap, also have a logging in persistence and give uninit compute a different logging name, but this will be the PR after.
in that notebook I also tried to create a live figure directly with the used dataset, thats what I had in mind with the comment from this morning.
in that notebook I also tried to get the metric out of the loop. not sure whether this is a useful way to go.

aaronspring commented:
as mentioned in the notebook: hist needs to be adapted in bootstrap, now different verification for historical skill possible

bradyrx commented:
On the notebook. It's right now just to produce the figures for the docs page, not for users to see since it's buried in `docs/source/images/alignment`. I can make a separate notebook in the future that goes into examples with a binder to play around with alignment. 

The `proplot` is not necessary but makes the docs figures significantly easier for me to make. It would be a pain to do in `matplotlib` with the subplots and everything. Again, this is just for figure docs, not for users to play with. I still think users direct experience should be with the log. The figures are just a visual aid I made manually.

bradyrx commented:
**NEXT PR** (in addition to maximize):

- Add "DEBUG" level logging with all inits/verifs printed out.

bradyrx commented:
Updated `compute_uninitialized` to actually get the right results for different alignments and added testing. Might break wherever we use it since it requires the hindcast as input now. Need to run to a meeting, will fix those breaks after.

aaronspring commented:
> Updated `compute_uninitialized` to actually get the right results for different alignments and added testing. Might break wherever we use it since it requires the hindcast as input now. Need to run to a meeting, will fix those breaks after.

I meant the historical alignment in computing uninitialized as it is used by the bootstrapping function.

aaronspring commented on climpred/bootstrap.py:
```diff
@@ -441,6 +443,7 @@ def bootstrap_compute(
             compute(
                 uninit_hind,
                 verif,
+                alignment=alignment,
```
Here I think the hist alignment could be implemented. But when the tests pass and function still works it’s ok for me. It will change in the uninit reference compute refactoring

bradyrx replied:
Yeah it's being handled here since it's still going through the main compute functions. When we refactor we'll implement the new method.

aaronspring replied:
so currently hist is not aligned. the refactoring PR then should handle this. Maybe hist can be an optional argument in the alignment function and then a union with hist/obs/hind should probably do it.
```
hind minmax 1954-01-01 00:00:00 2017-01-01 00:00:00
hist minmax 1955-01-01 00:00:00 2015-01-01 00:00:00
veri minmax 1948-01-01 00:00:00 2017-01-01 00:00:00

 same_verif
INFO:root:`compute_hindcast` for metric pearson_r and comparison e2o at 2020-03-04 22:10:05.002853
++++++++++++++++++++++++++++++++++++++++++++++++
INFO:root:lead=01 | dim=time | inits=1956-01-01 00:00:00-2016-01-01 00:00:00 | verif=1957-01-01 00:00:00-2017-01-01 00:00:00
INFO:root:lead=02 | dim=time | inits=1955-01-01 00:00:00-2015-01-01 00:00:00 | verif=1957-01-01 00:00:00-2017-01-01 00:00:00
INFO:root:lead=03 | dim=time | inits=1954-01-01 00:00:00-2014-01-01 00:00:00 | verif=1957-01-01 00:00:00-2017-01-01 00:00:00
INFO:root:`compute_hindcast` for metric pearson_r and comparison e2o at 2020-03-04 22:10:05.412592
++++++++++++++++++++++++++++++++++++++++++++++++
INFO:root:lead=01 | dim=time | inits=1957-01-01 00:00:00-2012-01-01 00:00:00 | verif=1958-01-01 00:00:00-2013-01-01 00:00:00
INFO:root:lead=02 | dim=time | inits=1956-01-01 00:00:00-2011-01-01 00:00:00 | verif=1958-01-01 00:00:00-2013-01-01 00:00:00
INFO:root:lead=03 | dim=time | inits=1955-01-01 00:00:00-2010-01-01 00:00:00 | verif=1958-01-01 00:00:00-2013-01-01 00:00:00
INFO:root:`compute_hindcast` for metric pearson_r and comparison e2o at 2020-03-04 22:10:05.798249
++++++++++++++++++++++++++++++++++++++++++++++++
INFO:root:lead=01 | dim=time | inits=1956-01-01 00:00:00-2016-01-01 00:00:00 | verif=1957-01-01 00:00:00-2017-01-01 00:00:00
INFO:root:lead=02 | dim=time | inits=1955-01-01 00:00:00-2015-01-01 00:00:00 | verif=1957-01-01 00:00:00-2017-01-01 00:00:00
INFO:root:lead=03 | dim=time | inits=1954-01-01 00:00:00-2014-01-01 00:00:00 | verif=1957-01-01 00:00:00-2017-01-01 00:00:00

bradyrx replied:
I see what you mean. Yeah I put a kludge into `compute_uninitialized` right now to find that union.
